broadinstitute/ddp-study-server
# [DSM] DDP-8349: lookup fields bug

## Context

Ticket: https://broadinstitute.atlassian.net/browse/DDP-8349

From very beginning when `PatchRoute` refactored, we wrote the script which was doing export on every input change, but problem arose for look up type fields, look up type fields send several `NameValue` objects and triggering elastic export for each them was losing data, since our script was way faster for reading than elastic for writing values into it, most simplest and cleanest approach was to send only one ES request for several `NameValue` object.

**Things have done:** 

- Changed `GeneratorPayload` to hold collection of `NameValue` instead of single `NameValue`.
- Overloaded method in `BasePatch` which takes list of `NameValue` if we have several name value pairs.
- Used template method design pattern for methods `getElement` and `parseElement` to avoid duplication and added condition in this methods if we have several name value objects to behave differently.
- Added tests for `CollectionSourceGenerator` and `CollectionMappingGenerator` to test generate method.

**How to test:**

1. Find the participant which has either medical record or onc history tab. (You can easily filter participants by choosing `Duplicate` column of `Medical Record` and search for `Dupliacte` - `Yes`).
2. If it is medical record tab, open it and choose any medical record, it will transfer you to the medical record page, where you will see the input called `Confirmed Institution Name`, put some letter in this field and wait look up functionality to find appropriate options for you, choose any of it, then go back, reload the participant list, visit the same participant and check if chosen values are presented correctly.
3. If it is onc history tab, open it and you will see one or several onc histories, you can directly put values into columns without opening it, find column called `Facility` and do the same look up procedure like medical record case.

## Checklist

- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :sweat_smile: There might be some issues here

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.

## Testing

- [ ] I have written zero automated tests but have poked around locally to verify proper functionality

## Release
